### PR TITLE
Convenience method to allow customizing route dependencies

### PR DESCRIPTION
**Pull request by alukach**
**Related Issue(s):** #

**Description:**

We're using stac-fastapi on a project and would like to add auth requirements to a number of endpoints.  To the best of my knowledge, there's not an official way to do this in this codebase.  This PR adds a convenience method (`add_route_dependencies`) to the main `StacApi` class.  This allows us to do something like the following:

```py
from fastapi import Depends, HTTPException, status, security
from stac_fastapi.api.app import StacApi

http_basic = security.HTTPBasic()

def must_be_bob(credentials: security. HTTPBasicCredentials = Depends(http_basic)):
    if credentials.username == 'bob':
        return True
    raise HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="You're not Bob",
        headers={"WWW-Authenticate": "Basic"},
    )

api = StacApi(
    route_dependencies=[
        # Only Bob can edit our data...
        (
            [
                {"path": "/collections", "method": "POST"},
                {"path": "/collections", "method": "PUT"},
                {"path": "/collections/{collectionId}", "method": "DELETE"},
                {"path": "/collections/{collectionId}/items", "method": "POST"},
                {"path": "/collections/{collectionId}/items", "method": "PUT"},
                {"path": "/collections/{collectionId}/items/{itemId}", "method": "DELETE"},
            ],
            [Depends(must_be_bob)]
        ),
    ]
)

# Or, alternatively
api = StacApi()
api.add_route_dependencies(
    scopes=[
        {"path": "/collections", "method": "POST"},
        {"path": "/collections", "method": "PUT"},
        {"path": "/collections/{collectionId}", "method": "DELETE"},
        {"path": "/collections/{collectionId}/items", "method": "POST"},
        {"path": "/collections/{collectionId}/items", "method": "PUT"},
        {"path": "/collections/{collectionId}/items/{itemId}", "method": "DELETE"},
    ],
    dependencies=[Depends(must_be_bob)]
)

```

This manner of injecting dependencies into already-existing routes doesn't seem well documented in either the FastAPI documentation or this codebase, so I figure it may be of use to someone.  It seems to work well and properly updates the OpenAPI spec as well:

image

I'm admittedly a little concerned that this is a bit heavy-handed (i.e. jamming more logic into that class), however wanted to present this solution as a starting point.  I'm all ears if there's a better way to achieve our goal.

### Update

To be clear about the status of this PR, I consider it still in the "proof-of-concept" stage.  The following should occur before merging:

- [x] Get sign-off that this is of interest
- [x] Get agreement on interface
- [x] Write tests

**PR Checklist:**

- [x] Code is formatted and linted (run `pre-commit run --all-files`)
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable, and docs build successfully (run `make docs`)
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/master/CHANGES.md).

### CODE REVIEW

File: stac_fastapi/api/stac_fastapi/api/routes.py [hunk 1]
```diff
@@ -94,3 +96,40 @@ def _endpoint(
...
+class Scope(TypedDict, total=False):
+    # More strict version of Starlette's Scope
+    # https://github.com/encode/starlette/blob/6af5c515e0a896cbf3f86ee043b88f6c24200bcf/starlette/types.py#L3
+    path: str
+    method: str
+    type: Optional[str]
+
+
+def add_route_dependencies(
+    routes: List[BaseRoute], scopes: List[Scope], dependencies=List[Depends]
```
Comment (vincentsarago) [inline stac_fastapi/api/stac_fastapi/api/routes.py:110]: `dependencies` should be of type `fastapi.params.Depends`

File: stac_fastapi/api/stac_fastapi/api/routes.py [hunk 2]
```diff
@@ -94,3 +96,40 @@ def _endpoint(
...
+        for route in routes:
+
+            match, _ = route.matches({"type": "http", **scope})
+            if match != Match.FULL:
+                continue
+
+            # Mimicking how APIRoute handles dependencies:
+            # https://github.com/tiangolo/fastapi/blob/1760da0efa55585c19835d81afa8ca386036c325/fastapi/routing.py#L408-L412
+            for depends in dependencies[::-1]:
+                route.dependant.dependencies.insert(
```
[the remark this answers is not in the document]
Comment (alukach) [reply stac_fastapi/api/stac_fastapi/api/routes.py:130]: Is this actually an issue?  `fastapi.routing.APIRoute` is a descendent of `starlette.routing.BaseRoute` ([`fastapi.routing.APIRoute`](https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/routing.py#L300) -> [`starlette.routing.Route`](https://github.com/encode/starlette/blob/881af38cc2e038de1f02e4fba7b7dab2df79b38a/starlette/routing.py#L185) -> [`starlette.routing.BaseRoute`](https://github.com/encode/starlette/blob/881af38cc2e038de1f02e4fba7b7dab2df79b38a/starlette/routing.py#L155)), does mypy not like that?

[the remark this answers is not in the document]
Comment (vincentsarago) [reply stac_fastapi/api/stac_fastapi/api/routes.py:131]: 🤷‍♂️ I just get Mypy warning in VSCode, maybe can ignore if the linter is happy

File: stac_fastapi/api/stac_fastapi/api/app.py [hunk 3]
```diff
@@ -312,6 +319,12 @@ async def ping():
 
         self.app.include_router(mgmt_router, tags=["Liveliness/Readiness"])
 
+    def add_route_dependencies(
+        self, scopes: List[Scope], dependencies=List[Depends]
```
[the remark this answers is not in the document]
Comment (alukach) [reply stac_fastapi/api/stac_fastapi/api/app.py:323]: This `Depends` is currently is imported from `fastapi.params`

[the remark this answers is not in the document]
Comment (vincentsarago) [reply stac_fastapi/api/stac_fastapi/api/app.py:323]: 🤦

**Review by vincentsarago — COMMENTED** [review]
👍 looks good to me! I've taken the liberty to add this to TiTiler as well 🙏 @alukach

**Review by geospatial-jeff — APPROVED** [review]
This is really cool!

**Comment by vincentsarago** [comment]
Can we add tests before merging 🙏

**Comment by alukach** [comment]
Sounds good.  It sounds like we're fine with the idea and interface, so I'll go ahead and write some tests and maybe build out the docstring a bit more for better documentation.

**Comment by AsgerPetersen** [comment]
This is very handy indeed! Looking forward to having it in main! Thank you all.

**Comment by jonhealy1** [comment]
Cool yea!

**Comment by vincentsarago** [comment]
@alukach do you think you could add tests and maybe some docs? It seems that people are interested having this merged in master 😄

**Comment by alukach** [comment]
Sorry about the delay all, I had forgotten about this PR 😬 (thanks for the ping, @vincentsarago!)

@vincentsarago @geospatial-jeff Added some tests in [146e28b](https://github.com/stac-utils/stac-fastapi/pull/295/commits/146e28b0f5a5846a0bfb13cccd84b748273fcdfd).  Can you run approve the workflow?  I wanted them to be separate from the SqlAlchemy and the PgStac tests.  To make them work, I had to create a dummy core client and dummy transaction client.  I'm open to suggestions on how these tests can be cleaner.

**Review by vincentsarago — APPROVED** [review]

**Comment by vincentsarago** [comment]
thanks @alukach 🙏

**Comment by alukach** [comment]
@AsgerPetersen @jonhealy1 If either of you have a chance to try this out, let me know how it goes!

**Comment by keul** [comment]
@alukach OK for authentication check (which is super cool) but… is it possible that this approach is not working for adding new query parameters to the defined route?

I mean: like in the FastAPI hello world example about dependencies: https://fastapi.tiangolo.com/tutorial/dependencies/#create-a-dependency-or-dependable

**Comment by alukach** [comment]
> is it possible that this approach is not working for adding new query parameters to the defined route?

@keul Can you share more about what you're trying to do?  There's a difference between having a dependency passed in as an argument to a FastAPI route (as you linked to) and dependencies inside a path operator ([link](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/)).  This PR does the later, it allows a dependency to be run before a route's logic is run, however you can't really do anything with the results.

**Comment by keul** [comment]
@alukach my need is to inject custom query parameters for the `/collections` endpoint (so, in the `all_collections` method).

As an example: what if I want to implement the collection pagination? https://github.com/radiantearth/stac-api-spec/tree/main/ogcapi-features#collection-pagination

But this is probably a topic for a new issue, sorry for the noise

**Comment by vincentsarago** [comment]
@keul I guess you'll need to change this https://github.com/stac-utils/stac-fastapi/blob/25879afe94296eb82b94b523bfa2871b686e035a/stac_fastapi/api/stac_fastapi/api/app.py#L214-L233 on your app to add your `page` dependency (instead of `EmptyRequest`)

**Comment by alukach** [comment]
Yes, I agree with what @vincentsarago has said.  

@keul When you say that you want to inject a custom query parameter, I suspect that you're leaving out the detail that you probably want to _do something_ with that custom query parameter later.  Assuming that is true, then you probably don't want to be augmenting an existing endpoint with the `add_route_dependencies()` convenience method and instead you probably want to be _overriding_ an existing endpoint with a custom endpoint handler that implements the logic that you need.

> But this is probably a topic for a new issue, sorry for the noise

Yeah, it may be best to open a [discussion](https://github.com/stac-utils/stac-fastapi/discussions) for something like this.  You're welcome to mention my username if you'd like to get my attention for further conversation.

**Pull request merged**